packages/lit-dev-content/site/articles/article/redux-reactive-controllers.md

```diff
@@ -0,0 +1,731 @@
+    State,
+    Action extends RAction,
+    Result = unknown
+> implements ReactiveController {
+  ...
+  selected: Result;
+
+  constructor(
+    host: ReactiveControllerHost,
+    store: Store<State, Action>,
+    selector: Selector<State, Result>,
+  ) {
+    ...
+    this.selected = selector(store.getState());
+  }
+}
+```
+
+### Updating the Component on Store Update
+
+We now have a controller that initializes to the initial state of the store, next let’s update `this.selected` when the state updates and then tell the host element to re-render when we have detected a change in the selected value.
+
+Redux stores have a [`Store.subscribe(listener)`](https://redux.js.org/api/store#subscribelistener) method which will call a given callback whenever the state of the store updates. Let's hook into this, update `this.selected`, and tell the host to update when the component is connected to the DOM:
+
+```ts
+...
+export class SelectorController<
+    State,
+    Action extends RAction,
+    Result = unknown
+> implements ReactiveController {
+  ...
+
+  hostConnected() {
+    this.store.subscribe(() => {
+      const selected = this.selector(this.store.getState());
+      this.selected = selected;
+      this.host.requestUpdate();
+    });
+  }
+}
+```
+
+Great! Now the controller will update its value and tell the host element to update whenever the state changes. A problem we may encounter here is that we will call `host.requestUpdate()` whenever *any* state changes and not specifically when our selected value changes. In this case, we should do an equality check and let the user decide if they would like to implement their own equality check:
+
+```ts
+...
+export type EqualityCheck = (a: unknown, b: unknown) => boolean;
+const tripleEquals: EqualityCheck = (a, b) => a === b;
+
+export class SelectorController<
+    State,
+    Action extends RAction,
+    Result = unknown
+> implements ReactiveController {
+  ...
+  equalityCheck : EqualityCheck;
+
+  constructor(
+    host: ReactiveControllerHost,
+    store: Store<State, Action>,
+    selector: Selector<State, Result>,
+    equalityCheck = tripleEquals
+  ) {
+    ...
+    this.equalityCheck = equalityCheck;
+  }
+
+  hostConnected() {
+    this.store.subscribe(() => {
+      const selected = this.selector(this.store.getState());
+      if (!this.equalityCheck(this.selected, selected)) {
+        this.selected = selected;
+        this.host.requestUpdate();
+      }
+    });
+  }
+}
+```
+
+By first comparing the previous state to the current state, we can avoid re-rendering components that don't need to be re-rendered which can improve performance. Nice!
+
+In conclusion, we need to improve our component so that it does not re-render when the component is disconnected from the page and the store’s state changes. Redux’s `Store.subscribe()` method returns an `unsubscribe()` function. Let’s keep track of this and unsubscribe from the store’s changes when the component disconnects.
```

Comment:
Logically it feels like unsubscribe should be implemented before the equality check as a subscribe/unsubscribe and connected/disconnected generally should be considered together. Equality check is an optimization that can be done afterwards.

Reply:
Also I'm debating whether configurable equality check is even needed. With redux recommending immutable patterns, triple equal should be good enough and it removes some complexity from the basic demonstration here.

Reply:
`===` would cause the array to re-render every state change in the shapes-list. a shallow equals would be useful here. Actually they offer a [shallowEqual](https://github.com/reduxjs/react-redux/blob/master/src/utils/shallowEqual.ts) in their react redux library

Reply:
For redux state, the array should only be a new reference if any items inside it has changed so `===` works. The `shallowEqual` in `react-redux` is used for comparing state/props across React component re-renders which would always be a new object.

I don't disagree that if we were to actually vend this as a package, we should make the equality check configurable. I wanted to raise whether it's worth the complication in a reactive controller demonstration.

Ultimately, I think it's fine to keep as an example of what things could be options for a controller. I do still think the order should be flipped with the disconnected.

Reply:
removed